### ISSUE TITLE
API: add start timesheet endpoint (make $begin optional)

### ISSUE THREAD

**Issue by amanzone** [kimai/kimai#1949]
In my case I would configure a kimai server for the company accessible on the WAN, and allow people to use terminal clients (e.g. Kimai2_cmd) to start and stop timesheets.

What I experienced is that when the timesheet is started by the employee the local PC time is used but when the timesheet is stopped the time used for the end tag is the server's time.

If the local PC (possibly non connected to the ADDC ntp) time is different from the kimai's server time, for instance by 2 minutes, then when I stop a timesheet the end time could be 13:08 (server's time). If then I immediatly start a new timesheet (on a different project for instance) then the local PC time would be used, for instance 13:06.

This means that I would have a timesheet for project A stopping at 13:08, and another timesheet for project B starting at 13:06 (while I was still working on project A as far as Kimai knows).

Does anyone know how to address this issue?

Thanks in advance for any help.

BR,
A.

**Comment by kevinpapst** [comment]
You general assumption is wrong, as Kimai uses always the server time. But the start time can obviously be changed by user depending on the time-tracking mode and the way you started the timesheet (restart button via form with editable start time).

Lets get @infeeeee on board, as he is the maintainer of the cmd.

**Comment by kevinpapst** [comment]
But having a look at the API the problem might be, that there is no explicit start method (only stop) and the create endpoint requires a begin date. Then start time has to be calculated via the cmd on client side and that might be different, as you noticed.

Setting up NTP is obviously the first thing to do. Then you could try to work with rounding rules as well.

**Comment by amanzone** [comment]
Yes, as far as I can see that is the issue: the "/api/timesheets" POST requires a "begin". If we could have the "begin" parameter optional, in case it is not provided the server may anyway use its own time and the issue would be solved.

Of course I may be oversimplifying the issue

Just my two cents...

**Comment by kevinpapst** [comment]
In theory its easy, in practice there are obstacles like the time tracking-modes.
I think I have to add an explicit endpoint for starting records.

**Comment by infeeeee** [comment]
I can add a workaround in kimai2-cmd: before the start a new timesheet, it asks the server for it's time, and uses that time on a new api call. I guess the time in the response header is the server time, so a simple ping would be enough.

But if you plan to add this option to the api, I wouldn't work on this, as this "bug" affects all clients.

**Comment by kevinpapst** [comment]
@infeeeee can you have a look at the linked PR #2007 - this is the first step, so you can fetch the users time (depending on its timezone). I still want to make a new endpoint, but well ... free time is currently rare ...

**Comment by infeeeee** [comment]
I implemented this, and works with the `api-user-now` branch: https://github.com/infeeeee/kimai2-cmd/releases/tag/1.2.0

I found that PATCH ​/api​/timesheets​/{id}​/restart also doesn't require a begin value, and always uses server time. So this bug shouldn't show up if users use restart instead of start, @amanzone if you don't want to update to this branch yet.

**Comment by kevinpapst** [comment]
Not sure when exactly this behavior changed, but I was able to create running timesheets by simply skipping the `begin` field. Even if the API docs marked it as required. For the next version, the field will be marked as optional.
